### PR TITLE
Move lint init functions

### PR DESCRIPTION
**Pull request by robplee**
So this is an attempt to address https://github.com/zmap/zlint/issues/371 .  Generally, it's not a particularly crazy set of changes.  For each of the lints in zlint I copied the init function to the top of the list of functions, right underneath the type definitions.  Furthermore, if there was a reference to an RFC/root policy document in the top comment, I've moved that to be right above the definition of the init function as that seemed to me to be what the goal of the issue was anyway!

I appreciate this will be fairly foul to review but at least the changes are simple?

### CODE REVIEW

**Comment by cardonator** [comment]
After just glancing through these changes it made me wonder if this actually improves the readability of the lints at all? This seems to be a difference of preference as the current model is something like "what the lint does, and then why the lint exists" vs. this change as more of "why the lint exists, and then what the lint does to enforce it".

I don't have a strong preference but I do understand the original design of tossing boilerplate metadata at the bottom of the file so it's easier to get to the meat of a Lint.

**Comment by sleevi** [comment]
Yes, this optimized a lot more for reading than writing; my assumption is every new lint coming in for review, and every existing lint that someone may be reading (e.g. to figure out why it's firing) benefits from having the "why" before the "how", in the spirit of https://chris.beams.io/posts/git-commit/

It does mean when  you fire up a lint, if you're familiar with the why, you spend more time getting to the how. But I think that seems less likely/rarer, in practice, compared to trying to understand and/or contribute a new one?

**Comment by cpu** [comment]
I'm swayed by @sleevi's arguments on this. I also think having the name and source that the lint is registered under at the top of the file is best because that's what a consumer would need to include/exclude the lint.

**Comment by cpu** [comment]
> I appreciate this will be fairly foul to review but at least the changes are simple?

I can own doing a first pass of this if nobody else is available however I won't have a chance until Sunday.

**Comment by cardonator** [comment]
I was easily swayed 😄 Thanks for the link @sleevi!

**Comment by sleevi** [comment]
I'll take a stab at reviewing. In the process, I've noted a few out-dated citations or inconsistent preambles: do we want to fix them in this (e.g. by adding suggestions to batch), or do we want to defer them? I'm slightly inclined to fix now, since they're only textual (comments and citation), but I can be happily swayed otherwise.

**Comment by cardonator** [comment]
I would say let's fix them now since this is already touching so many files in those exact areas. That would be my suggestion.

**Comment by cpu** [comment]
> I'll take a stab at reviewing

Thanks! I'll kick the assignment over to you for now.

> I would say let's fix them now

SGTM too

**Review by sleevi — COMMENTED** [review]
Woo! This was a big slog, but huge thanks for doing this!

I tried to limit my suggestions to those I could use inline suggest for, so that you could batch accept them. I'd suggest holding off until Monday to doing so, just so @cpu and @cardonator can chime in if they have any concerns with the suggestions. Most of them are around the BRs, and just updating references or text to reflect changes. I filed several bugs as a result of this, so thanks for doing this, it helps get an extra set of eyes on things!

**Comment by christopher-henderson** [comment]
Woof! :stuck_out_tongue: 

I'm kind of interested to see what kind of tooling you built to pull this off (...I assume you didn't do this by hand?). Perhaps we might be able to use whatever you wrote to help verify all this.

Perhaps put this through the [Go parser](https://golang.org/pkg/go/parser/)? One would think (hope) that we get the same AST before and after this change (unless the order of declaration changes the AST in some way...I hope not. Although these small string changes will muck with it a bit).

Otherwise, I admit that I tried to spot check this visually and instantly found myself taking mental shortcuts.

**Comment by cpu** [comment]
Apologies, I merged https://github.com/zmap/zlint/pull/532 before I thought of this WIP PR. If possible could you fix `v3/lints/cabf_ev/lint_ev_organization_id_missing.go` in this branch as well?

File: v3/lints/cabf_br/lint_cab_dv_conflicts_with_street.go
Comment (robplee) [reply v3/lints/cabf_br/lint_cab_dv_conflicts_with_street.go:39]: A quick question as there seem to be a bunch of these '.1'->'.4' changes, is this a case of the citation in the code being out of date and not reflecting changes made to the cabf br document since this lint was written or were they wrong to start with and are finally being fixed now?

File: v3/lints/cabf_br/lint_ext_san_contains_reserved_ip.go
Comment (robplee) [reply v3/lints/cabf_br/lint_ext_san_contains_reserved_ip.go:24]: Lines have been deleted in commit 065c11a

File: v3/lints/cabf_br/lint_sub_cert_cert_policy_empty.go
Comment (robplee) [reply v3/lints/cabf_br/lint_sub_cert_cert_policy_empty.go:29]: Lines have been deleted in commit 065c11a

**Comment by robplee** [comment]
Well.  That was a bit of a slog.  Big thanks to Sleevi for the review and all the suggestions.  I think I've addressed everything raised (including refactoring Daniel's new lint) so I think, build permitting, that this is done from my side.

Christopher-Henderson, I'd love to say that I did something clever to do all the function moving but that wouldn't be true, I just manually clicked and dragged through!  That being said, thanks for the suggestions made by Sleevi, there are some 'functional changes' which could potentially cause changes in the result from the go parser because there are some strings changed here and there which I'd assume would cause problems but please correct me if I'm wrong!

Thanks again for all the review efforts.

File: v3/lints/cabf_br/lint_cab_dv_conflicts_with_street.go
Comment (sleevi) [reply v3/lints/cabf_br/lint_cab_dv_conflicts_with_street.go]: They were changed in https://cabforum.org/2020/09/09/ballot-sc35-cleanups-and-clarifications/ in response to CA confusion about the intersection of the requirements and subscriber certs vs CA certs.

Comment (sleevi) [reply v3/lints/cabf_br/lint_cab_dv_conflicts_with_street.go]: Oh, and the existing lints already checked if they were not-CA certs, so the lints were correctly scoped to only subscriber certs, but the requirement was moved from 7.1.6.1 (all certs) to 7.1.6.4 (subscriber certs)

**Review by sleevi — APPROVED** [review]
Thanks! Looks good to me!

**Comment by zakird** [comment]
Thank you both, @robplee and @sleevi for doing this. I'll merge in a bit once tests pass.

**Comment by cpu** [comment]
Seems like the integration test status check reporting might be jammed, I see the [action did run and pass](https://github.com/zmap/zlint/actions/runs/483325081).

**Review by christopher-henderson — APPROVED** [review]
I gave it a quick go at scoping out parsing this diff, but even without the string changes the [ast types](https://golang.org/pkg/go/ast/#Package) don't appear to have any `Equal` methods, so I just don't think scripting this is in the cards.

So I gave it another glance over and at the very least my eyes couldn't spot any irregularities.

**Review by cpu — APPROVED** [review]
Thanks again @robplee ! I'll fix up the lint I merged in https://github.com/zmap/zlint/pull/533/files separately as a follow-up in a moment.

**Comment by cpu** [comment]
> I gave it a quick go at scoping out parsing this diff, but even without the string changes the ast types don't appear to have any Equal methods, so I just don't think scripting this is in the cards.

@christopher-henderson I'd love to see a CI task that fails the build if new lints declare `init` in the wrong order in the AST to keep ourselves honest moving forward. Do you think that's feasible given your experimentation so far?

**Comment by christopher-henderson** [comment]
@cpu 

Oh now _that_ would be fun! Lemme check that out today (off the top of my head it should be feasible as the AST should be deterministic). Gonna track that in #546

**Pull request merged**
